File: CppSolution/ReverseNodesink-Group.cpp
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution 
{
private:
	ListNode *reverse(ListNode *pre, ListNode *nxt)
	{

        ListNode *last = pre->next;
        ListNode *cur = last->next;
        while (cur != nxt)
        {
            last->next = cur->next;
            cur->next = pre->next;
            pre->next = cur;
            cur = last->next;
        }
        return last;
	}
public:
    ListNode *reverseKGroup(ListNode *head, int k) 
    {
        if (head == NULL || k == 1) return head;
        ListNode *dummy = new ListNode(0);
        dummy->next = head;
        ListNode *pre = dummy;
        int i = 0;
        while (head != NULL)
        {
            i++;
            if (i % k == 0)
            {
                pre = reverse(pre, head->next);
                head = pre->next;
            }
            else head = head->next;
        } 
        return dummy->next;
    }
};

int main()
{
	
	return 0;
}

File: JavaSolution/src/FaceBook/ReverseNodesinkGroup.java
package FaceBook;

/**
 * Created by xyunpeng on 1/5/16.
 */
public class ReverseNodesinkGroup {
    public ListNode reverseKGroup(ListNode head, int k) {
        if (k == 0 || head == null || head.next == null) {
            return head;
        }

        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode pre = dummy;
        ListNode cur = head;
        int count = 0;
        while (cur != null) {
            count++;
            ListNode next = cur.next;
            if (count == k) {
                count = 0;
                pre = reverseKElements(pre, next);
            }
            cur = next;
        }

        return dummy.next;
    }

    private ListNode reverseKElements(ListNode head, ListNode tail) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode pre = head.next;
        ListNode cur = head.next.next;
        while (cur != tail) {
            ListNode next = cur.next;
            cur.next = head.next;
            head.next = cur;
            cur = next;
        }
        pre.next = tail;
        return pre;
    }
}
